src/Api/referrals.js

```diff
@@ -342,34 +342,50 @@ export async function getReferralCodeOwner(chainId, referralCode) {
   const referralStorageAddress = getContract(chainId, "ReferralStorage");
   const provider = getProvider(null, chainId);
   const contract = new ethers.Contract(referralStorageAddress, ReferralStorage.abi, provider);
-  const codeOwner = await contract.codeOwners(referralCode);
-  return codeOwner;
+  try {
+    const codeOwner = await contract.codeOwners(referralCode);
+    return codeOwner;
+  } catch (error) {
+    console.error(error);
+  }
 }
 
-export function useUserReferralCode(library, chainId, account) {
-  const referralStorageAddress = getContract(chainId, "ReferralStorage");
-
-  let { data: userReferralCode, mutate: mutateUserReferralCode } = useSWR(
-    account && [`ReferralStorage:traderReferralCodes`, chainId, referralStorageAddress, "traderReferralCodes", account],
-    {
-      fetcher: fetcher(library, ReferralStorage),
-    }
-  );
-
+export function useUserReferralCode(chainId, account) {
+  const [userReferralCode, setUserReferralCode] = useState(null);
+  const [userReferralCodeForExchange, setUserReferralCodeForExchange] = useState(ethers.constants.HashZero);
+  const [userReferralCodeString, setUserReferralCodeString] = useState("");
   const userReferralCodeInLocalStorage = window.localStorage.getItem(REFERRAL_CODE_KEY);
-  if (isHashZero(userReferralCode) && userReferralCodeInLocalStorage) {
-    userReferralCode = userReferralCodeInLocalStorage;
-  }
+  const contract = useMemo(() => {
+    const referralStorageAddress = getContract(chainId, "ReferralStorage");
+    const provider = getProvider(null, chainId);
+    return new ethers.Contract(referralStorageAddress, ReferralStorage.abi, provider);
+  }, [chainId]);
 
-  let userReferralCodeString;
-  if (userReferralCode && !isHashZero(userReferralCode)) {
-    userReferralCodeString = decodeReferralCode(userReferralCode);
-  }
+  useEffect(() => {
+    async function getUserReferralCode() {
+      if (account) {
+        const referralCode = await contract.traderReferralCodes(account);
+        if (!isHashZero(referralCode)) {
+          setUserReferralCode(referralCode);
+          setUserReferralCodeString(decodeReferralCode(referralCode));
+          setUserReferralCodeForExchange(ethers.constants.HashZero);
+        } else if (userReferralCodeInLocalStorage && userReferralCodeInLocalStorage.startsWith("0x")) {
+          const localstorageCodeOwner = await contract.codeOwners(userReferralCodeInLocalStorage);
+          if (!isAddressZero(localstorageCodeOwner)) {
+            setUserReferralCode(userReferralCodeInLocalStorage);
+            setUserReferralCodeForExchange(userReferralCodeInLocalStorage);
+            setUserReferralCodeString(decodeReferralCode(userReferralCodeInLocalStorage));
+          }
+        }
+      }
+    }
+    getUserReferralCode();
+  }, [account, contract, userReferralCodeInLocalStorage]);
 
   return {
     userReferralCode,
     userReferralCodeString,
-    mutateUserReferralCode,
+    userReferralCodeForExchange,
```

Comment:
i think better to return boolean flag like "isCodeAttached" or something like this. with this flag other app components can understand if the code is already on-chain or not. and SwapBox can decide if it wants to use it inside `increasePosition` call

imo with `userReferralCodeForExchange` API lib knows too much about it's "user" – SwapBox